nep-monitor.mediawiki

```diff
@@ -0,0 +1,171 @@
+* Developers & investors assessing network suitability for their business case
+* Developers looking to improve network performance
+* Token holders looking to make an educated vote for consensus node candidates
+* (Future) Token holders to build or maintain trust in the network
+* Node hosters looking for hosting requirements and monitoring tools
+* Media outlets comparing blockchain platforms
+
+Current data attribute collection happens for most part via RPC calls which requires RPC server availability (enabled and port exposed). RPC calls only allow for collecting a single data attribute per call therefore requiring multiple RPC calls to collect all attributes. This pull based mechanism will put unnecessary load on the node when collecting data and repeats for every client requesting this data. If RPC server capabilities are not supported in the particular client, enabled in the configuration, or ports are not properly forwarded when supported, valuable node statistics are lost. Therefore, we feel a new monitoring interface and data attribute set that keeps the node load to a minimum while expanding insights is justified.
+
+
+==Specification==
+
+The specification consists of three parts 
+# architectural changes for exposing the monitor interface in a load efficient manner 
+# a description of the <code>monitor</code> payload format 
+# a description of the new data attribute set.
+
+===Monitor interface===
+The monitor interface is the access point where collection clients can access the data attributes. When a client connects to a node over the P2P or Websocket port a handshake protocol is started. First, the node sends a [http://docs.neo.org/en-us/network/network-protocol.html#network-message network message] with a <code>version</code> command and the appropriate payload. In the original handshake implementation the client is expected to respond with a similar <code>version</code> command network message and its own payload or be disconnected from the node. According to this proposal, we allow the client to provide an alternative response in the form of a network message holding the <code>monitor</code> command type and a zero length payload to indicate that the client is not a regular peer but a client that wishes to collect data attributes. The node separates the connection into its own monitoring channel on which no [http://docs.neo.org/en-us/network/network-protocol.html#network-message regular network messages] from the client are accepted and only data from the node is broadcasted. The <code>monitor</code> payload format is described in the next section. 
+
+Data attribute broadcasting is to occur after every block persist. In the [https://github.com/neo-project/neo current official C# client] this means on every <code>Blockchain.PersistCompleted</code> event. A visual representation of the idea can be seen below
```

Comment:
Thanks for the input!

regarding _stream mode_. I don't think that stream mode is ideal in this proposal. As we all know, blocks do not get generated in exact equal intervals, while the majority of the current default attributes only change per block. So then it becomes a trade-off; 
- fixed interval with a majority data duplication
- variable/triggered interval with no/minimal data duplication

regarding _push mode_,  what interval is the data pushed with?

takeaways:
2. that might be an implementation caveat worth noting. 
3. I think we need to do some profiling on this to see if pushing "more bytes per client" (=no request profile) performs better/worse than "custom data per client" (=with request profile). So we need to look at performance of 
- collect data + craft 1 `Message` then distribute * _n_ peers
- collect data + craft 1 `Message` * _m_ unique profiles , then distribute to peers
I have a feeling that unique profiles will be slower because there's logic that needs to filter data per profile, each profile `Message` requires a new double SHA256 checksum, and there's logic for distributing the specific message to the specific client. I'm still open for it, hoping to get input from others as well!